ToolChain/Assembler/Assembler.py

```diff
@@ -56,79 +58,109 @@ def __init__(self, inputFile=None, outputFile=None):
+        globalSymbols = []                   # Temporary list to hold the names of global(external) symbols
+        localSymbols = []                    # Temporary list to hold the names of local(internal) symbols
+        externalSymbols = []                 # Master list of all external symbols to be used by linker
+        internalSymbols = []                 # Master list of all internal symbols
+        offset = 0                           # Holds the instruction's length, used to calculate offset
+        lineno = 0                           # Keeps track of which line we're parsing to pinpoint errors
+        relativeAddressCounter = 0           # Used to determine memory address of a label at a given point in the file
+
+        try:
+            # Initial test to verify that the file we're reading actually exists
+            file = open(inputFile, mode="r")
+            fileLines = file.readlines()
+            file.close()
+        except KeyError as e:
+            raise ValueError("Error - unable to open file: " + inputFile)
+
+        for line in fileLines:
+            lineno += 1
+            # Here we begin our evaluation of each line. We get the binary instruction, offset, and label flag
+            # Memory locations are relative to the beginning of the file. Instructions and data are given set lengths.
+            # Internal and external symbol lists contain tuples containing the symbol name, and its offset.
+            # These will be used to fill the .o file information before the binary data.
+
+            try:
+                # This will catch any exception thrown by the parser. If there is one, we print the error,
+                # The line at which it occurred, and we quit out.
+                instruction, offset, labelFlag = parser.parse(line)
+                relativeAddressCounter += offset
+            except ValueError as e:
+                print(str(e) + " at line: {}".format(lineno))
+                quit()
+
+            if labelFlag == INSTRUCTION_FLAG:
+                # If the label flag is set to 0, it's a regular instruction, string, number, of memref.
+                binaryDataString += instruction
+
+            elif labelFlag == LOCAL_REFERENCE_FLAG:
+                # If the label flag is set to 1, it's an internal symbol. We ensure labels are unique when declared
+                if instruction not in localSymbols:
+                    localSymbols.append(instruction)
+                    labelTuple = (instruction, relativeAddressCounter)
+                    internalSymbols.append(labelTuple)
+                else:
+                    raise ValueError("Label already present in local list, cannot declare duplicate label name")
+                if instruction in globalSymbols:
+                    externalSymbols.append(labelTuple)
+
+            elif labelFlag == GLOBAL_REFERENCE_FLAG:
```

Comment:
If a symbol is declared before it is marked as global, the the symbol will never make it to the external symbol list. A linker error will probably follow when the used tries to link their application. This can be resolved right here, also preventing the user some time trying to figure out the linker error.